contracts/Staking.sol

```diff
@@ -0,0 +1,487 @@
+import "./ILockManager.sol";
+
+import "./Checkpointing.sol";
+
+import "@aragon/os/contracts/common/IsContract.sol";
+import "@aragon/os/contracts/common/TimeHelpers.sol";
+import "@aragon/os/contracts/lib/math/SafeMath.sol";
+import "@aragon/os/contracts/lib/token/ERC20.sol";
+
+
+contract Staking is ERCStaking, ERCStakingHistory, TimeHelpers, IsContract {
+    using SafeMath for uint256;
+    using Checkpointing for Checkpointing.History;
+
+    uint64 private constant MAX_UINT64 = uint64(-1);
+
+    string private constant ERROR_NOT_LOCK_MANAGER = "STAKING_NOT_LOCK_MANAGER";
+    string private constant ERROR_AMOUNT_ZERO = "STAKING_AMOUNT_ZERO";
+    string private constant ERROR_TOKEN_TRANSFER = "STAKING_TOKEN_TRANSFER";
+    string private constant ERROR_NOT_ENOUGH_BALANCE = "STAKING_NOT_ENOUGH_BALANCE";
+    string private constant ERROR_CAN_NOT_UNLOCK = "STAKING_CAN_NOT_UNLOCKERROR_";
+    string private constant ERROR_INVALID_LOCK_ID = "STAKING_INVALID_LOCK_ID";
+    string private constant ERROR_UNLOCKED_LOCK = "STAKING_UNLOCKED_LOCK";
+    string private constant ERROR_INCREASING_LOCK_AMOUNT = "STAKING_INCREASING_LOCK_AMOUNT";
+
+    struct Lock {
+        uint256 amount;
+        uint64 unlockedAt;
+        ILockManager manager; // can also be an EOA
+        bytes data;
+    }
+
+    struct Account {
+        uint256[] activeLockIds;
+        uint256 lastLockId;
+        mapping(uint256 => Lock) locks; // first valid lock starts at 1, so _toLockId = 0 means no lock
+    }
+
+    ERC20 stakingToken;
+    mapping(address => Account) accounts;
+    mapping (address => Checkpointing.History) stakedHistory;
+    Checkpointing.History totalStakedHistory;
+
+    event Staked(address indexed account, uint256 amount, uint256 total, bytes data);
+    event Unstaked(address indexed account, uint256 amount, uint256 total, bytes data);
+    event Locked(address indexed account, uint256 indexed lockId, uint256 amount, address manager, bytes data);
+    event Unlocked(address indexed account, uint256 indexed lockId, uint256 amount, address manager, bytes data);
+    event LockAmountChanged(address indexed account, uint256 indexed lockId, uint256 amount);
+    event LockManagerChanged(address indexed account, uint256 indexed lockId, address manager);
+    event LockDataChanged(address indexed account, uint256 indexed lockId, bytes data);
+    event Transferred(address indexed from, uint256 indexed fromLockId, uint256 amount, address to, uint256 toLockId);
+
+    modifier isLockManager(address _account, uint256 _lockId) {
+        require(
+            msg.sender == address(accounts[_account].locks[_lockId].manager),
+            ERROR_NOT_LOCK_MANAGER
+        );
+        _;
+    }
+
+    constructor(ERC20 _stakingToken) public {
+        require(isContract(_stakingToken));
+        stakingToken = _stakingToken;
+    }
+
+    /* External functions */
+
+    /**
+     * @notice Stakes `_amount` tokens, transferring them from caller
```

Comment:
Well, it's already using Radspec, I guess you mean it can be improved. I didn't include `data` because i don't think it would be helpful. It wouldn't be human readable, it would be hard to translate (as each lock manager can have a very different format) and it could make the description unnecessarily long.